### DISCUSSION TITLE
How to configure Metals to prevent OutOfMemoryError?

### DISCUSSION THREAD

**Discussion by adamw** [scalameta/metals#6785] [Q&A] [answered]
In my recent Tapir coding session I got:

```
2024.09.18 16:09:08 INFO  time: compiled zioHttpServer in 0.86s
2024.09.18 16:14:34 INFO  time: code lens generation in 4.53s

Exception: java.lang.OutOfMemoryError thrown from the UncaughtExceptionHandler in thread "pool-1-thread-369"
[Error - 16:15:05] Request textDocument/codeAction failed.
  Message: Internal error.
  Code: -32603 
java.lang.OutOfMemoryError: Java heap space
```

As the project is quite large, I suspect I'd need to allocate more memory for Metals, but I'm not sure (a) what's the default, (b) which setting is the correct one.

In the extension settings I found "Ammonite Jvm Properties" and "Bloop Jvm Properties", though I'm not sure if it's one of those. I'd suspect the Bloop one, though I didn't yet read the article you linked in the other question so I'm still not quite sure what Bloop is ;)

Anyway, having a setting like "Metals memory" or "Metals Jvm Properties" (as it's Metals that I'm using as an IDE) would be great. And to have some mention as to what is the default, so that I could bump that to *2.

**Comment by tgodzik** [comment, marked as answer]
You can use serverProperties:
```
  "metals.serverProperties": [
    "-Xmx2G",
    "-Dmetals.loglevel=debug",
    "-Dmetals.statistics=all",
    "-Dmetals.enable-best-effort=true",
    "-XX:+UseZGC",
    "-XX:ZUncommitDelay=30",
    "-XX:ZCollectionInterval=5"
  ],
```

the above is my setting. With UseZGCand JDK 17 you might even skip -Xmx. The default is only 1g.

**Reply by adamw** [reply]
Thanks, though where do I put this? Are these some other settings than under Manage -> Settings -> Extensions -> Metals?

**Reply by tgodzik** [reply]
these are normal settings in VS Code, the page actually corresponds to a json file. You can also look for `metals.serverProperties` in the settings and add them manually.

**Reply by adamw** [reply]
Ah found it. Thanks! :)
